lib/internal/repl/await.js

```diff
@@ -51,24 +53,63 @@ const visitorsWithoutAncestors = {
     walk.base.ReturnStatement(node, state, c);
   },
   VariableDeclaration(node, state, c) {
-    if (node.kind === 'var' ||
-        state.ancestors[state.ancestors.length - 2] === state.body) {
-      if (node.declarations.length === 1) {
-        state.replace(node.start, node.start + node.kind.length, 'void');
-      } else {
-        state.replace(node.start, node.start + node.kind.length, 'void (');
-      }
+    if (node.declarations.length === 1) {
+      state.replace(node.start, node.start + node.kind.length, 'void');
+    } else {
+      state.replace(node.start, node.start + node.kind.length, 'void (');
+    }
+
+    ArrayPrototypeForEach(node.declarations, (decl) => {
+      state.prepend(decl, '(');
+      state.append(decl, decl.init ? ')' : '=undefined)');
+    });
 
-      ArrayPrototypeForEach(node.declarations, (decl) => {
-        state.prepend(decl, '(');
-        state.append(decl, decl.init ? ')' : '=undefined)');
-      });
+    if (node.declarations.length !== 1) {
+      state.append(node.declarations[node.declarations.length - 1], ')');
+    }
+
+    const declarationSeparator = ', ';
+
+    function getVariableDeclarationIdentifier(node) {
+      switch (node.type) {
+        case 'Identifier':
+          return node.name;
+        case 'ObjectPattern':
+          return ArrayPrototypeJoin(
+            ArrayPrototypeMap(
+              node.properties,
+              (property) => getVariableDeclarationIdentifier(property.value)
+            ),
+            declarationSeparator
+          );
+        case 'ArrayPattern':
+          return ArrayPrototypeJoin(
+            ArrayPrototypeMap(
+              node.elements,
+              (element) => getVariableDeclarationIdentifier(element)
+            ),
+            declarationSeparator
+          );
+      }
+    }
 
-      if (node.declarations.length !== 1) {
-        state.append(node.declarations[node.declarations.length - 1], ')');
+    const variableIdentifiersToHoist = [];
+    for (const decl of node.declarations) {
+      const identifier = getVariableDeclarationIdentifier(decl.id);
+      if (identifier !== undefined) {
+        variableIdentifiersToHoist.push(identifier);
       }
     }
```

Comment:
Avoid `for-of` on arrays as it is at risk of prototype pollution.
```suggestion
    const variableIdentifiersToHoist = ArrayPrototypeFlatMap(
      node.declarations,
      (decl) => getVariableDeclarationIdentifier(decl.id) ?? []
    );
```
Alternatively we can use a `for(;;)` loop.

Reply:
I've simplified the registering process so this shouldn't be an issue anymore.